external/wakaama/core/er-coap-13/er-coap-13.c

```diff
@@ -71,1446 +70,1533 @@ const char *coap_error_message = "";
+
+	PRINTF("-Done serializing at %p----\n", option);
+
+	coap_pkt->options_len = option - options_buf;
+
+	PRINTF("- Serialized options_len : %d\n", coap_pkt->options_len);
+
+	uint32_t len = coap_pkt->options_len;
+
+	if (coap_pkt->payload_len > 0) {
+		len += COAP_OPTIONS_MARKER_LEN + coap_pkt->payload_len;
+	}
+
+	PRINTF("Options+Payload (len %u)", len);
+	if (len < 13) {
+		coap_pkt->buffer[0] = (COAP_HEADER_TOKEN_LEN_MASK & coap_pkt->token_len);
+		coap_pkt->buffer[0] |= (len << 4);
+		coap_pkt->buffer[1] = coap_pkt->code;
+		option = coap_pkt->buffer + 2;
+	} else if (len < ((1 << 8) + 13)) {
+		coap_pkt->buffer[0] = (COAP_HEADER_TOKEN_LEN_MASK & coap_pkt->token_len);
+		coap_pkt->buffer[0] |= (13 << 4);
+		coap_pkt->buffer[1] = len - 13;
+		coap_pkt->buffer[2] = coap_pkt->code;
+		option = coap_pkt->buffer + 3;
+	} else if (len < ((1 << 16) + 269)) {
+		coap_pkt->buffer[0] = (COAP_HEADER_TOKEN_LEN_MASK & coap_pkt->token_len);
+		coap_pkt->buffer[0] |= (14 << 4);
+		coap_pkt->buffer[1] = ((len - 269) & 0xFF00) >> 8;
+		coap_pkt->buffer[2] = (len - 269) & 0x00FF;
+		coap_pkt->buffer[3] = coap_pkt->code;
+		option = coap_pkt->buffer + 4;
+	} else {
+		coap_pkt->buffer[0] = (COAP_HEADER_TOKEN_LEN_MASK & coap_pkt->token_len);
+		coap_pkt->buffer[0] |= (15 << 4);
+		coap_pkt->buffer[1] = ((len - 65805) & 0xFF000000) >> 24;
+		coap_pkt->buffer[2] = ((len - 65805) & 0x00FF0000) >> 16;
+		coap_pkt->buffer[3] = ((len - 65805) & 0x0000FF00) >> 8;
+		coap_pkt->buffer[4] = (len - 65805) & 0x00FF;
+		coap_pkt->buffer[5] = coap_pkt->code;
+		option = coap_pkt->buffer + 6;
+	}
+
+	/* set Token */
+	PRINTF("Token (len %u)", coap_pkt->token_len);
+
+	for (current_number = 0; current_number < coap_pkt->token_len; ++current_number) {
+		PRINTF(" %02X", coap_pkt->token[current_number]);
+		*option = coap_pkt->token[current_number];
+		++option;
+	}
+	PRINTF("-\n");
+
+	/* Copy options */
+	memcpy(option, options_buf, coap_pkt->options_len);
+	option += coap_pkt->options_len;
+	lwm2m_free(options_buf);
+
+	/* Free allocated header fields */
+	coap_free_header(packet);
+
+	/* Pack payload */
+	/* Payload marker */
+	if (coap_pkt->payload_len) {
+		*option = 0xFF;
+		++option;
+	}
+
+	memmove(option, coap_pkt->payload, coap_pkt->payload_len);
+
+	PRINTF("-Done %u B (header len %u, payload len %u)-\n", coap_pkt->payload_len + option - buffer, option - buffer, coap_pkt->payload_len);
+
+	PRINTF("Dump [0x%02X %02X %02X %02X  %02X %02X %02X %02X]\n", coap_pkt->buffer[0], coap_pkt->buffer[1], coap_pkt->buffer[2], coap_pkt->buffer[3], coap_pkt->buffer[4], coap_pkt->buffer[5], coap_pkt->buffer[6], coap_pkt->buffer[7]
+		  );
+
+	return (option - buffer) + coap_pkt->payload_len;	/* packet length */
 }
 
 /*-----------------------------------------------------------------------------------*/
-size_t
-coap_serialize_message(void *packet, uint8_t *buffer)
+size_t coap_serialize_message(void *packet, uint8_t *buffer)
```

Comment:
Could you please explain the benefit from having to separate tcp and udp based coap serialization? Wouldn't it be tedious to check for protocol everytime application developer calls coap serialize function?